lib/puppet/parser/functions/lastpass_note_read.rb

```diff
@@ -0,0 +1,127 @@
+require 'fileutils'
+require 'yaml'
+require 'English'
+# require 'etc'
+
+# Retrieves from a LastPass secure note
+#
+# Useful for having data that is managed via LastPass and shared with Puppet.
+#
+# Usage: lastpass_note_read(folder, name)
+# Example: $db_config = lastpass_note_read('oracle/db', 'appuser')
+Puppet::Parser::Functions.newfunction(:lastpass_note_read, :type => :rvalue) do |args|
+  raise Puppet::ParseError, 'Usage: lastpass_note_read(folder, name)' unless args.size == 2
+
+  folder = args[0]
+  raise Puppet::ParseError, 'Must provide folder' if folder.empty?
+
+  name = args[1]
+  raise Puppet::ParseError, 'Must provide data name' if name.empty?
+
+  username = lookupvar('lastpass::username')
+
+  cmd = `which lpass`
+  raise Puppet::ParseError, 'lpass command not found' if cmd.empty?
+  version = `lpass --version`.strip
+  raise Puppet::ParseError, "unexpected lpass version: #{version}" unless version =~ /v1.1.2$/
+
+  # TODO: consider how sync should be handled
+
+  status = `lpass status`.strip
+
+  if $CHILD_STATUS.exitstatus != 0
+    raise Puppet::ParseError, "error: lpass status: #{status}" unless status == 'Not logged in.'
+    login_result = `lpass login #{username}`
+    raise Puppet::ParseError, "error: lpass login: #{login_result}" unless login_result =~ /^Success:/
+  end
+
+  ls_result = `lpass ls '#{folder}'`
+  if $CHILD_STATUS.exitstatus != 0
+    raise Puppet::ParseError, "error: lpass ls '#{folder}': #{status}"
+  end
+
+  ls_result =~ %r{#{Regexp.escape(folder)}/#{Regexp.escape(name)} \[id: ([^\]]+)\]}
+  id = Regexp.last_match(1)
+  unless id
+    raise Puppet::ParseError, "error: unable to find id for '#{folder}/#{name}' in :\n#{ls_result}"
+  end
+
+  show_result = `lpass show '#{id}'`
+  if $CHILD_STATUS.exitstatus != 0
+    raise Puppet::ParseError, "error: lpass show '#{folder}/#{name}' [id: #{id}]: #{status}"
+  end
+
+  # Parsing the output could use some cleaning up, but this works for now.
+  # The output appears to have reasonably consistent structure.
+  # The first line is the SHARE/GROUP\PATH/UNIQUENAME [id: nnnn]
+  # Following lines are Field Name: ...
+  # The Notes field seems to be last, which makes sense since it can be
+  # multi-line. Once it starts, just read everything else into it.
```

Comment:
Yeah. This is ugly. I'm really disappointed that the LastPass CLI can't be told to format the response in something useful. You can query individual fields instead of getting it all at once, so it would be possible to query the type and then fetch each field you want. You get just the value that way. But that would mean several shell executions and we'll be doing this every time Puppet runs.

I was meaning to ask if Puppet custom functions can broken down into smaller functions. For example, I'd like to have a function to do the login and then call it from the main function. Same with parsing the response and other actions. That way they can be reusable. Is the only way to do that to create more Puppet custom functions? Is that acceptable?